swupd/delta.go

```diff
@@ -31,149 +43,123 @@ func CreateDeltas(manifest, statedir string, from, to uint32) ([]*File, error) {
-				if err != nil {
-					failedChan <- f
-				}
+			for delta := range deltaQueue {
+				delta.Error = createDelta(c, delta)
 			}
-		}(statedir)
+		}()
 	}
 
-	// Send jobs to the queue for delta goroutines to pick up
-	for _, file := range deltas {
-		deltaQueue <- file
+	// Send jobs to the queue for delta goroutines to pick up.
+	for i := range deltas {
+		deltaQueue <- &deltas[i]
 	}
 
 	// Send message that no more jobs are being sent
 	close(deltaQueue)
 	wg.Wait()
 
-	// Once wait finishes we can signal the collector that no more jobs exist
-	close(failedChan)
-
-	return <-failedList, nil
+	return deltas, nil
 }
 
-func consolidateDeltaFiles(manifest *Manifest, from uint32, c config) ([]*File, error) {
-	var files []*File
+func createDelta(c *config, delta *Delta) error {
+	oldPath := filepath.Join(c.imageBase, fmt.Sprint(delta.From.Version), "full", delta.From.Name)
+	newPath := filepath.Join(c.imageBase, fmt.Sprint(delta.To.Version), "full", delta.To.Name)
+	deltaDir := filepath.Join(c.outputDir, fmt.Sprint(delta.To.Version), "delta")
+	deltaName := fmt.Sprintf("%d-%d-%s-%s", delta.From.Version, delta.To.Version, delta.From.Hash, delta.To.Hash)
+	deltaPath := filepath.Join(deltaDir, deltaName)
 
-	if manifest == nil {
-		return nil, nil
+	if err := helpers.RunCommandSilent("bsdiff", oldPath, newPath, deltaPath); err != nil {
+		_ = os.Remove(deltaPath) // Might have returned FULLDL
+		return errors.Wrapf(err, "Failed to create delta for %s -> %s", oldPath, newPath)
 	}
 
-	for _, file := range manifest.Files {
-		if file.Version <= from {
-			continue
-		}
-		if file.DeltaPeer == nil {
-			continue
-		}
-		if file.Type != TypeFile ||
-			file.DeltaPeer.Type != TypeFile {
-			continue
-		}
-
-		deltadir := filepath.Join(c.outputDir, fmt.Sprintf("%d", file.Version), "delta")
-		fromToString := fmt.Sprintf("%d-%d-%s-%s", file.DeltaPeer.Version, file.Version, file.DeltaPeer.Hash, file.Hash)
-
-		deltaFile := filepath.Join(deltadir, fromToString)
-
-		// Only add to list if the delta file does not already exist on the system
-		if _, err := os.Stat(deltaFile); os.IsNotExist(err) {
-			// Only create deltas for files bigger than 200 bytes
-			fullname := filepath.Join(c.imageBase, fmt.Sprintf("%d", file.Version), "full", file.Name)
-			file.Info, _ = os.Stat(fullname)
-			if file.Info.Size() > 200 {
-				files = append(files, file)
-			}
-		}
-
+	// Check that the delta actually applies correctly.
+	testPath := filepath.Join(deltaDir, "."+deltaName+".testnewfile")
+	if err := helpers.RunCommandSilent("bspatch", oldPath, testPath, deltaPath); err != nil {
+		return errors.Wrapf(err, "Failed to apply delta %s", deltaPath)
 	}
-
-	return files, nil
-}
-
-func createDelta(c config, file *File, statedir string, fromVersion uint32, fromHash Hashval) error {
-	// File without peers cannot have deltas
-	if file.DeltaPeer == nil {
-		return nil
+	testHash, err := Hashcalc(testPath)
```

Comment:
If the above check passed we know testPath exists on the filesystem. The test file should be removed even in the error cases so `defer os.Remove(testPath)` before this line. I don't see a benefit of keeping it around in the error cases.

Reply:
If Hashcalc fails, we don't know for sure that the file exists.

I agree that we should remove it regardless. This was just the logic from the previous existing code, I'll change it.